hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/defrag/SnapshotDefragService.java

```diff
@@ -135,48 +196,318 @@ private boolean isRocksToolsNativeLibAvailable() {
+      Table<String, byte[]> checkpointTable = checkpointDBStore.getTable(table, StringCodec.get(),
+          ByteArrayCodec.get());
+      String tableBucketPrefix = prefixInfo.getTablePrefix(table);
+      String prefixUpperBound = getLexicographicallyHigherString(tableBucketPrefix);
+
+      Pair<String, String> tableBounds = getTableBounds(checkpointTable);
+      String tableLowestValue = tableBounds.getLeft();
+      String tableHighestValue = tableBounds.getRight();
+
+      // If lowest value is not null and if the bucket prefix corresponding to the table is greater than lower then
+      // delete the range between lowest value and bucket prefix.
+      if (tableLowestValue != null && tableLowestValue.compareTo(tableBucketPrefix) < 0) {
+        checkpointTable.deleteRange(tableLowestValue, tableBucketPrefix);
+      }
+      // If highest value is not null and if the next higher lexicographical string of bucket prefix corresponding to
+      // the table is less than equal to the highest value then delete the range between bucket prefix
+      // and also the highest value.
+      if (tableHighestValue != null && tableHighestValue.compareTo(prefixUpperBound) >= 0) {
+        checkpointTable.deleteRange(prefixUpperBound, tableHighestValue);
+        checkpointTable.delete(tableHighestValue);
+      }
+      // Compact the table completely with kForce to get rid of tombstones.
+      try (ManagedCompactRangeOptions compactRangeOptions = new ManagedCompactRangeOptions()) {
+        compactRangeOptions.setBottommostLevelCompaction(ManagedCompactRangeOptions.BottommostLevelCompaction.kForce);
+        compactRangeOptions.setExclusiveManualCompaction(true);
+        checkpointDBStore.compactTable(table, compactRangeOptions);
+      }
+    }
+  }
 
-    // TODO: Implement full defragmentation
+  /**
+   * Spills table difference into an SST file based on the provided delta file paths,
+   * current snapshot table, previous snapshot table, and an optional table key prefix.
+   *
+   * The method reads the delta files and compares the records against the snapshot tables.
+   * Any differences, including tombstones (deleted entries), are written to a new SST file.
+   *
+   * @param deltaFilePaths the list of paths to the delta files to process
+   * @param snapshotTable the current snapshot table for comparison
+   * @param previousSnapshotTable the previous snapshot table for comparison
+   * @param tableKeyPrefix the prefix for filtering certain keys, or null if all keys are to be included
+   * @return a pair of the path of the created SST file containing the differences and a boolean
+   *         indicating whether any delta entries were written (true if there are differences, false otherwise)
+   * @throws IOException if an I/O error occurs during processing
+   */
+  @VisibleForTesting
+  Pair<Path, Boolean> spillTableDiffIntoSstFile(List<Path> deltaFilePaths, Table<String, byte[]> snapshotTable,
```

Comment:
We could move this and other helper methods to a separate static helper class to reduce length of SDS class